Update ink feature splashFactory test coverage to not rely on the Theme's default value

willlockwood opened the pull request:
## Description

This is the first in a series of PRs planned for several small improvements to the `InkWell` and its ripple animations. See https://github.com/flutter/flutter/issues/73163 for additional context:

> SplashFactory Test Coverage: Currently, many tests fail after changing the default splash factory on the material Theme. To make the PR changing the default splash factory easier, test coverage should be improved for widgets using ink features, to be able to pass regardless of what the theme’s default splash factory is. We can take the opportunity here to consolidate some duplicated code for utils related to testing how ripples are drawn.

This PR is just updating tests for widgets using `InkWell`s, to include test coverage for when the `InkWell`s in question will use both the `InkRipple.splashFactory` and the `InkSplash.splashFactory`. Specifically, this PR is meant to improve test coverage, so that when we update the default `splashFactory` on the material `Theme` from the `InkSplash.splashFactory` to the `InkRipple.splashFactory`, it isn't necessary to update a bunch of failing tests, but just add a small test for the default value on the `Theme`.

Also, you'll find that I added some small utils in `ink_paint_test_utils.dart`, to help consolidate a bunch of boilerplate code. This contributed to some extra diffs. Similarly, in places where it made sense, I organized tests into groups, which similarly led to a bunch of extra diffs.

## Related Issues

- [#73163: [proposal] InkRipple and InkSplash implementations warrant improvements](https://github.com/flutter/flutter/issues/73163)

## Tests

I added/modified the following tests:
`bottom_navigation_bar_test`:
- **BottomNavigationBar shifting backgroundColor with transition**

`chip_test`
- _Ink features
  - with InkSplashFactory_
    - **Chip creates centered, unique splash when label is tapped**
    - **Delete button creates non-centered, unique splash when tapped**
    - **Chip without delete button creates correct ripple**
    - **Selection with avatar works as expected on RawChip**
    - **Selection without avatar works as expected on RawChip**
    - **Activation works as expected on RawChip**
  - _with InkRippleFactory_
    - **Chip creates centered, unique splash when label is tapped**
    - **Delete button creates non-centered, unique splash when tapped**
    - **Chip without delete button creates correct ripple**
    - **Selection with avatar works as expected on RawChip**
    - **Selection without avatar works as expected on RawChip**
    - **Activation works as expected on RawChip**

`data_table_test`
- _DataRow custom colors_
  - **DataRow renders custom colors when selected**
  - **DataRow renders custom colors when disabled**
  - **DataRow with InkSplash splash factory renders custom colors when pressed**
  - **DataRow with InkRipple splash factory renders custom colors when pressed**

`icon_button_test`
- _Ink features_
  - **IconButton with explicit splashColor and highlightColor**
  - _with InkSplash splash factory_
    - **IconButton draws splash with explicit splashRadius**
  - _with InkRipple splash factory_
    - **IconButton ripple radius is 5 greater than splashRadius**

`ink_paint_test`
- **The InkWell widget renders an ink splash**
- **Does the Ink widget render anything**
- **The InkWell widget renders an SelectAction or ActivateAction-induced ink ripple**
- **Canceling InkRipple that was disposed when its animation ended draws one transparent circle**

`ink_well_test` (a bunch of organizing)
- _Ink feature color states_
  - _on hover_
    - **InkWell uses hoverColor**
    - **InkWell uses resolved overlayColor MaterialState.hovered**
  - _on focus_
    - **ink response changes color on focus**
    - **ink response changes color on focus with overlayColor**
    - **ink response uses radius for focus highlight**
    - **ink response doesn't change color on focus when on touch device**
  - _on pressed_
    - _with InkSplash splash factory_
      - **splash color matches splashColor**
      - **splash color matches resolved overlayColor for MaterialState.pressed**
    - _with InkSplash splash factory_
      - **ripple color matches splashColor**
      - **ripple color matches resolved overlayColor for MaterialState.pressed**

`raw_material_button_test`
- _Ink Features_
  - _with InkSplash.splashFactory_
    - **RawMaterialButton responds with splash when tapped**
    - **RawMaterialButton responds to shortcut with splash when activated**
    - **Ink splash from center tap originates in correct location**
    - **Ink splash from tap above material originates in correct location**
  - _with InkRipple.splashFactory_
    - **RawMaterialButton responds with ripple when tapped**
    - **RawMaterialButton responds to shortcut with ripple when activated**
    - **Ink ripple from center tap originates in correct location**
    - **Ink ripple from tap above material originates in correct location**

`tabs_test`
- **Tab's InkSplash color matches resolved Tab overlayColor for MaterialState.pressed**
- **Tab's InkRipple color matches resolved Tab overlayColor for MaterialState.pressed**

*Replace this with a list of the tests that you added as part of this PR. A change in behavior with no test covering it
will likely get reverted accidentally sooner or later. PRs must include tests for all changed/updated/fixed behaviors. See [Test Coverage].*

## Checklist

Before you create this PR, confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*

<!-- Links -->
[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

willlockwood committed:
Consolidate ripple pattern matcher and use it in the ink paint test, and fix test without explicit splashFactory

willlockwood committed:
Fix tabs_test.dart by adding explicit ink feature cases for each splash factory

willlockwood committed:
Reorganize/fix raw_material_button_test.dart by adding explicit cases for each splash factory

willlockwood committed:
Reorganize/fix icon_button_test.dart by adding explicit cases for each splash factory

willlockwood committed:
Reorganize/fix chip_test.dart by adding explicit cases for each splash factory

willlockwood committed:
Reorganize data table tests to include ink splashes and consolidate boilerplate

willlockwood committed:
Organize/fix ink_well_test.dart with explicit cases for each splash factory

willlockwood committed:
Add explicit splashFactory to theme data in bottom_navigation_bar_test.dart so tests pass

willlockwood committed:
Remove color resolving boilerplate in ink_well_test.dart

willlockwood committed:
Don't use extensions for paint utils

willlockwood committed:
Remove anonymous param in data_table_test.dart onSelectChanged

willlockwood committed:
Fix control body lint in tabs_test.dart

willlockwood committed:
Try to fix linux analyze by removing extra spaces

willlockwood commented on packages/flutter/test/material/bottom_navigation_bar_test.dart:
```diff
@@ -1555,6 +1555,7 @@ void main() {
     Widget runTest() {
```
Reviewers: I'd recommend first looking at the additions in `ink_paint_test_utils`, because they're used in several tests

willlockwood commented on packages/flutter/test/material/chip_test.dart:
```diff
@@ -216,43 +223,6 @@ Widget _chipWithOptionalDeleteButton({
   );
 }
 
-bool offsetsAreClose(Offset a, Offset b) => (a - b).distance < 1.0;
-bool radiiAreClose(double a, double b) => (a - b).abs() < 1.0;
-
-// Ripple pattern matches if there exists at least one ripple
-// with the [expectedCenter] and [expectedRadius].
-// This ensures the existence of a ripple.
-PaintPattern ripplePattern(Offset expectedCenter, double expectedRadius) {
-  return paints
-    ..something((Symbol method, List<dynamic> arguments) {
-        if (method != #drawCircle)
-          return false;
-        final Offset center = arguments[0] as Offset;
-        final double radius = arguments[1] as double;
-        return offsetsAreClose(center, expectedCenter) && radiiAreClose(radius, expectedRadius);
-      }
-    );
-}
-
-// Unique ripple pattern matches if there does not exist ripples
-// other than ones with the [expectedCenter] and [expectedRadius].
-// This ensures the nonexistence of two different ripples.
-PaintPattern uniqueRipplePattern(Offset expectedCenter, double expectedRadius) {
-  return paints
-    ..everything((Symbol method, List<dynamic> arguments) {
-        if (method != #drawCircle)
-          return true;
-        final Offset center = arguments[0] as Offset;
-        final double radius = arguments[1] as double;
-        if (offsetsAreClose(center, expectedCenter) && radiiAreClose(radius, expectedRadius))
-          return true;
-        throw '''
-              Expected: center == $expectedCenter, radius == $expectedRadius
-              Found: center == $center radius == $radius''';
-      }
-    );
-}
-
```
A maximally robust version of this, with better debugging, is now in `ink_paint_test_utils`, and should be referenced for any ripple pattern paint testing

willlockwood committed:
Revert accidental reformat from dartFmt in data_table_test.dart

willlockwood commented on packages/flutter/test/material/ink_well_test.dart:
```diff
@@ -2,15 +2,16 @@
 // Use of this source code is governed by a BSD-style license that can be
```
Before:
<img width="350" src="https://user-images.githubusercontent.com/7915388/103831138-41f29000-5030-11eb-99e0-37245a57d880.png">

willlockwood commented:
@HansMuller if it helps, right now I can provide some pictures of the test structure before and after that hopefully can make the scope of the changes here clearer. Sorry for the large amount of diffs! If you'd like me to roll back anything to cut down the scope here just lemme know

willlockwood commented on packages/flutter/test/material/tabs_test.dart:
```diff
@@ -14,6 +14,7 @@ import '../rendering/mock_canvas.dart';
 import '../rendering/recording_canvas.dart';
```
Before and after: 
<img width="500" src="https://user-images.githubusercontent.com/7915388/103940821-d9b0b680-50e2-11eb-9773-522c93849203.png">
<img width="653" alt="Screen Shot 2021-01-07 at 12 24 16 PM" src="https://user-images.githubusercontent.com/7915388/103941157-4fb51d80-50e3-11eb-9ff1-c5493c341a6b.png">

willlockwood commented on packages/flutter/test/material/raw_material_button_test.dart:
```diff
@@ -11,97 +11,356 @@ import 'package:flutter_test/flutter_test.dart';
 
```
`raw_material_button_test`

Before:
<img width="500" alt="rmb_before" src="https://user-images.githubusercontent.com/7915388/103942445-3dd47a00-50e5-11eb-98c4-6680fba30199.png">

<img width="500" alt="rmb_after" src="https://user-images.githubusercontent.com/7915388/103942449-3f9e3d80-50e5-11eb-8a67-69a5e2d94b5a.png">

nt4f04uNd replied:
Why would we need tests for both InkRipple and InkSplash?

willlockwood commented on packages/flutter/test/material/ink_well_test.dart:
```diff
@@ -2,15 +2,16 @@
 // Use of this source code is governed by a BSD-style license that can be
 // found in the LICENSE file.
 
```
`ink_well_test`

Before:
<img width="500" alt="ink_well_before" src="https://user-images.githubusercontent.com/7915388/103943320-bdaf1400-50e6-11eb-8111-a01918476f3c.png">

After:
<img width="500" alt="ink_well_after" src="https://user-images.githubusercontent.com/7915388/103943323-bee04100-50e6-11eb-92c3-f5c247077c75.png">

willlockwood commented on packages/flutter/test/material/ink_paint_test.dart:
```diff
@@ -8,6 +8,7 @@ import 'package:flutter/services.dart';
 import 'package:flutter_test/flutter_test.dart';
 
 import '../rendering/mock_canvas.dart';
```
`ink_paint_test`

Only substantive change here is on line 58: adding a splashFactory to a test that shouldn't need duplication for each splashFactory, but currently depends on the default value on the theme

willlockwood commented on packages/flutter/test/material/icon_button_test.dart:
```diff
@@ -11,6 +11,7 @@ import 'package:flutter_test/flutter_test.dart';
 import '../rendering/mock_canvas.dart';
```
Before:
<img width="400" alt="ib_before" src="https://user-images.githubusercontent.com/7915388/103944676-e33d1d00-50e8-11eb-8422-ca3200da5419.png">

After
<img width="400" alt="ib_after" src="https://user-images.githubusercontent.com/7915388/103944679-e46e4a00-50e8-11eb-80b3-f7ee4c0186ef.png">

willlockwood commented on packages/flutter/test/material/data_table_test.dart:
```diff
@@ -12,6 +12,7 @@ import 'package:vector_math/vector_math_64.dart' show Matrix3;
 
 import '../rendering/mock_canvas.dart';
```
Before and after:
<img width="400" alt="dt_before" src="https://user-images.githubusercontent.com/7915388/103946328-4e87ee80-50eb-11eb-8084-c2e18491749b.png">
<img width="500" alt="dt_after" src="https://user-images.githubusercontent.com/7915388/103946330-4fb91b80-50eb-11eb-9568-fb0962c67601.png">

willlockwood commented on packages/flutter/test/material/chip_test.dart:
```diff
@@ -14,6 +14,7 @@ import 'package:flutter/scheduler.dart';
 import '../rendering/mock_canvas.dart';
 import '../widgets/semantics_tester.dart';
 import 'feedback_tester.dart';
```
Before and after:
<img width="450" alt="chips_before" src="https://user-images.githubusercontent.com/7915388/103947231-a115da80-50ec-11eb-950b-77b771e5fc24.png">

<img width="450" alt="chips_after" src="https://user-images.githubusercontent.com/7915388/103947234-a2470780-50ec-11eb-92f7-eaa7650e4f8e.png">

willlockwood commented:
Hey @HansMuller—no rush or anything, but just wanted to check in on this. Is there anything I can do to help make review for this any easier, or anything like that?

RA80533 commented:
/cc @HansMuller

nt4f04uNd commented:
@willlockwood could you please resolve merge conflict?

willlockwood commented:
> @willlockwood could you please resolve merge conflict?

Ah yes, will do today! Thanks for the nudge 👍🏻

willlockwood committed:
Replace Container with SizedBox for whitespace

willlockwood committed:
Fix tests to look for SizedBox instead of Container

willlockwood commented:
@HansMuller @nt4f04uNd 

OK cool this should be all set! Merge conflicts were just due to commits that changed some `Container`s to `SizedBox`es. I also ran all tests on this branch after changing the default value for a `ThemeData`'s `splashFactory`, and they're all passing, so the next PR updating the default should be real simple.

The point of this PR is just to make future changes (see https://github.com/flutter/flutter/issues/73163 for a full list) easier to make/test etc., so I'd really appreciate having the review whenever you get the chance, so I can get started on the rest of this stuff! 🙏🏻

nt4f04uNd commented on packages/flutter/test/material/bottom_navigation_bar_test.dart:
```diff
@@ -1582,6 +1582,7 @@ void main() {
     Widget runTest() {
       int _currentIndex = 0;
       return MaterialApp(
+        theme: ThemeData(splashFactory: InkSplash.splashFactory),
```
I suppose you are doing this temporarily for tests to pass?
In proposal you mention:

> Default SplashFactory: As a quick follow-up to the first PR, we should also change the default splash factory to the InkRipple.splashFactory, which is going to match the material animation for all Android APIs going forward. Currently, InkSplash.splashFactory is the default, which is an outdated choice, that's getting more outdated by the day.

After this change, all similar replacements of `InkSplash.splashFactory` in tests should be removed

willlockwood replied:
Ah yes good point—I remember doing this, because at the time, I didn't understand how to update/add goldens (which is how we're testing this widget at the moment).

I'll go through and add another set of tests for the bottom navigation bar, that uses a new set of goldens, so that eventually, if we wanted to delete any tests related to the InkSplash.splashFactory, that could be a quick deletion 👍🏻

nt4f04uNd commented on packages/flutter/test/material/ink_paint_test_utils.dart:
```diff
@@ -0,0 +1,61 @@
+  };
+}
+
+PaintPattern paintsRipple({Offset? tapDown, Offset? center, double? radius, Color? color, int? alpha, bool unique = false}) {
+  final PaintPattern pattern = paints;
+  if (tapDown != null) {
+    pattern.translate(x: 0, y: 0);
+    pattern.translate(x: tapDown.dx, y: tapDown.dy);
+  }
+  final PaintPatternPredicate predicate = _ripplePatternPredicate(center, radius, color, alpha, unique);
+  unique ? pattern.everything(predicate) : pattern.something(predicate);
+  return pattern;
+}
+
+MaterialInkController getMaterial<T>(WidgetTester tester) {
+  return Material.of(tester.element(find.byType(T)))!;
+}
+
+RenderObject getInkFeatures(WidgetTester tester) {
+  return tester.allRenderObjects.firstWhere((RenderObject object) => object.runtimeType.toString() == '_RenderInkFeatures');
+}
```
These would need a documentation I think

nt4f04uNd replied:
Also `getMaterial` is defined locally in some tests, so perhaps we should name it more carefully or rename those local ones

willlockwood replied:
Great points. I'll:
- Add some documentation, and try to follow the conventions that exist on the existing paint predicate methods on the tester, and
- Update the name of `getMaterialInkController`, which is more accurate to the actual return type (instances of `getMaterial` return an actual `Material`, so it would feel wrong to edit those instead), and would be a completely unique name for a method call.

nt4f04uNd commented on packages/flutter/test/material/chip_test.dart:
```diff
@@ -1017,417 +987,750 @@ void main() {
 
-    // Waits for 100 ms again.
-    await tester.pump(const Duration(milliseconds: 100));
+        expect(box, paintsRipple(center: const Offset(163, 6), radius: 42));
+        expect(box, paintsRipple(center: const Offset(163, 6), radius: 42, unique: true));
```
I'm wondering if we could remove this double-check here (and in other similar places) and leave only the unique one.
Am I missing some reason it's checked twice with unique and non-unique?

RA80533 replied:
If `unique` defaults to `false`, proper coverage would test both cases.

willlockwood replied:
Good question—I'd be happy to remove it if you'd prefer. Largely, I just saw that the previous test was testing for a ripple, and then for a unique ripple, and I thought it would be imprudent to arbitrarily take away expectations on a test that I'm refactoring, when I'm not actually changing the expected behavior related to the uniqueness?

willlockwood committed:
Add goldens for each splash factory for bottom_navigation_bar_test.dart

skia-gold commented:
Gold has detected about 48 untriaged digest(s) on patchset 17.
View them at https://flutter-gold.skia.org/cl/github/73391

willlockwood committed:
Update name of getMaterial to getMaterialInkController

nt4f04uNd commented:
Why do we need tests for both InkRipple and InkSplash? We need to be sure that Ink features are applied and eother of them will sufficient I guess

willlockwood committed:
Add documentation for new methods in ink_paint_test_utils.dart

willlockwood committed:
Delete getInkFeatures because we can actually always use getMaterialInkController instead

willlockwood commented:
> Why do we need tests for both InkRipple and InkSplash? We need to be sure that Ink features are applied and either of them will sufficient I guess

I guess I just figured, that many of the tests are testing specifically for things like the radius, offset, and alpha of the ripples, and these are details that pass tests specific to different splash factories, so it felt strange to just change the values used in those tests to reflect the new splash factory.

Would you prefer that I delete all the tests involving the InkSplash splash factory? I guess I also supposed, if they were superfluous, we could always delete them later as well, but I guess I just wasn't sure what the usual practice would be in case like this, where we're refactoring test cases. Happy to go in that direction if you'd like!

skia-gold commented:
Gold has detected about 32 untriaged digest(s) on patchset 20.
View them at https://flutter-gold.skia.org/cl/github/73391

nt4f04uNd commented:
> Also, another thought I had was that I'm planning on updating the actual implementations for the InkRipple and InkSplash in future PRs—would it be clearer/easier then, to update these test methods?

I think that they there should be independent tests for them, currently they are not tested separately, only as parts of other tests (InkSplash is tested, but it probably needs more). Then utils used to test them should be exposed for other tests to use them, so as you said, we wouldn't change them over all the places they are used

> Would you prefer that I delete all the tests involving the InkSplash splash factory?

I personally would. Note, that I'm just a regular contributor as you, I'm not part of the team

Friendly ping @HansMuller - could you please give it a review and say what do you think?

Piinks commented:
@willlockwood would you like to continue with this change? It looks like there are merge conflicts that will need to be resolved. Pushing a fresh commit should also regenerate the golden files that have expired. Are visual differences expected for this change?

willlockwood commented:
Yup, can do ASAP! Apologies for the delay 👍🏻

willlockwood committed:
Fix some stuff from the rebase

skia-gold commented:
Gold has detected about 21 new digest(s) on patchset 21.
View them at https://flutter-gold.skia.org/cl/github/73391

willlockwood commented:
Tests are all passing locally, and are also passing when I update the default value of `ThemeData` to the `InkRipple.splashFactory` (making it so that tests are passing regardless of which `splashFactory` is the default in `ThemeData` is kinda the point of this PR).

So, should be all set again for some eyes! Thanks @Piinks for the reminder (Are you in a position to review this? Or should I still wait for @HansMuller?). 

And to answer your question on the goldens, there are no UI changes in the framework associated with this PR, _**but**_, there are going to be out-of-date goldens because I've updated the names of a few existing golden tests and added a few more (the goal is, where there is test coverage for the ripple animations, I wanted to make sure there were golden tests for each `splashFactory`).

Piinks commented:
> Are you in a position to review this? Or should I still wait for @HansMuller?

I will defer to @HansMuller, I am not familiar with this area of the codebase

> And to answer your question on the goldens, there are no UI changes in the framework associated with this PR, _**but**_, there are going to be some golden diffs because I've updated the names of a few existing golden tests and added a few more (the goal is, where there is test coverage for the ripple animations, I wanted to make sure there were golden tests for each `splashFactory`).

I don't think we should change the names of existing golden files, that breaks the history of images we have for the given test. I would recommend leaving those names the same.

Also, it looks like this change needs to be rebased with the master branch.

Piinks commented:
(Triage) @willlockwood would you like to update this PR with the latest from master and restore the golden file names?

willlockwood commented:
> (Triage) @willlockwood would you like to update this PR with the latest from master and restore the golden file names?

Thank you for the reminder—will do as soon as I can!

And also, I'm kinda new to OSS stuff in general, and am now wondering, do y'all think I should just close this and start with somewhere user-facing? My goal has been to be able to make my way through a bunch of fidelity improvements to the ripple animation (see: [the related issue](https://github.com/flutter/flutter/issues/73163)), and I'm starting to think this may have been a bad way to start lol—this is how I'd do it at my work (start with PRs that will make future PRs easier to read/review) but I'm thinking maybe this is not the way to get eyes on contributions to larger open-source projects?

Btw I'm actually curious, so I hope there's no bitterness or anything like that, that y'all read in there—I'm definitely grateful for all of the people who have given any time to looking at this PR or the issue; I just want to figure out what would be the most effective approach for stuff like this :D

Piinks commented:
Wow! That is awesome. First off, thanks for planning out all of this work, and starting it! If you would like to discuss maybe the overall design and gather feedback that way first, I would recommend creating a design document. Especially if this is the first of several changes. We have [a template available here](https://docs.google.com/document/d/1SFRO8U2toOlAaZ38dsuEU7Wm5fn41wvBCWKiwADqfmw/edit?usp=sharing).
That will make it much easier to plan changes, migrations, and give code reviewers better context of the whole picture. WDYT?

willlockwood commented:
> ... I would recommend creating a design document. Especially if this is the first of several changes. We have [a template available here](https://docs.google.com/document/d/1SFRO8U2toOlAaZ38dsuEU7Wm5fn41wvBCWKiwADqfmw/edit?usp=sharing). That will make it much easier to plan changes, migrations, and give code reviewers better context of the whole picture. WDYT?

@Piinks Got it—sounds good. I finally have some time to come back to this now, so I'll go ahead and do that; thank you for the suggestion! I'm gonna try to gather some information about how the `splashFactory`/`InkWell` implementations are likely is going to be impacted by the Material 3 migration, and write out the doc for the broader set of fidelity improvements I imagine will be useful as we continue to support (I think) the old ripple animations. Closing this for now until we can be more certain about what would be best for first steps.

Piinks commented:
Thanks @willlockwood! I look forward to seeing your doc. You may also want to reach out to @darrenaustin, he has been working on the Material 3 migration, and his work may intersect with yours. Cheers!

LasseRosenow commented:
I think this comment from @darrenaustin is what pink is talking about:

https://github.com/flutter/flutter/pull/94599#issuecomment-997979851

Hope this helps :)

willlockwood commented:
> I think this comment from darrenaustin is what Piinks is talking about:

Huh, I might be totally missing it, but from what I can tell, that discussion is mainly a back and forth between Tesser and Muller? 

And cc @darrenaustin—not sure what the best way is for me to pick your brain on some of the Material 3 splash-related stuff, so I added you on the discord and am hoping that might be convenient for ya? TL;DR: I'm trying to [improve the pre-Android 12 splash animations](https://github.com/flutter/flutter/issues/73163) and I wanna make sure that what I'm planning to do isn't going to be redundant, irrelevant, or counterproductive given what y'all are working on

HansMuller commented:
@willlockwood - it turns out the @clocksmith has been working on the Material 3 ink splash effect. I will ask him to sync-up with you here.

clocksmith commented:
Thank you for taking this on! I have some planned changes as well, but it looks like many are already covered in https://github.com/flutter/flutter/issues/73163. I will add my notes there.

FYI I also have a pending PR for the m3 ripple which uses shader. It would be great to add tests to that too.
